server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/ParquetFileAccessor.java

```diff
@@ -429,92 +427,174 @@ private void createParquetWriter() throws IOException, InterruptedException {
     /**
      * Generate parquet schema using schema file
      */
-    private MessageType readSchemaFile(String schemaFile)
+    private MessageType readSchemaFile(String schemaFile, List<ColumnDescriptor> columns)
             throws IOException {
         LOG.debug("{}-{}: Using parquet schema from given schema file {}", context.getTransactionId(),
                 context.getSegmentId(), schemaFile);
+        MessageType schema;
         try (InputStream inputStream = fs.open(new Path(schemaFile))) {
-            return MessageTypeParser.parseMessageType(IOUtils.toString(inputStream, StandardCharsets.UTF_8));
+            schema = MessageTypeParser.parseMessageType(IOUtils.toString(inputStream, StandardCharsets.UTF_8));
         }
+
+        validateParsedSchema(schema, columns);
+        return schema;
     }
 
     /**
-     * Generate parquet schema using column descriptors
+     * Validate parquet schema parsed from user provided schema file
+     *
+     * @param schema  the schema parsed from user provided schema file
+     * @param columns contains Greenplum column type information of a row
+     */
+    private void validateParsedSchema(MessageType schema, List<ColumnDescriptor> columns) {
+        if (schema.getFieldCount() != columns.size()) {
+            LOG.warn(String.format("Schema field count %s doesn't match column count %s", schema.getFieldCount(), columns.size()));
+        }
+
+        for (int i = 0; i < schema.getFieldCount(); i++) {
+            Type type = schema.getType(i);
+            if (!type.isPrimitive()) {
+                validateComplexType(type.asGroupType());
+            }
+            // TODO: Need to check ordering and type matching between schema and columns
+        }
+    }
```

Comment:
It's good you switched for iterating over schema and not GPDB columns for now. However, we still want to do iterator based traversal, not index-based access, so:
```suggestion
        for (Type type : schema.getFields()) {
            if (!type.isPrimitive()) {
                validateComplexType(type.asGroupType());
            }
            // TODO: Need to check ordering and type matching between schema and columns
        }
    }
```